krm-functions/interface-fn/design.md

```diff
@@ -0,0 +1,157 @@
+-------------------------------------------------------
+
+For each `Interface` the function will generate one `IPAllocation` and one `VLANAllocation` CR. In order to generate the `Spec` of those child resources it uses information also from the `ClusterContext` resource that the kpt package assumed to also contain.
+
+Let's see an example! Assuming that a kpt package contains the following two resources (among others):
+
+    apiVersion: req.nephio.org/v1alpha1
+    kind: Interface
+    metadata:
+        name: n6
+        annotations:
+            config.kubernetes.io/local-config: "true"
+    spec:
+        networkInstance:
+            name: vpc-internet
+        cniType: sriov
+        attachementType: vlan
+    ---
+    apiVersion: infra.nephio.org/v1alpha1
+    kind: ClusterContext
+    metadata: 
+        name: cluster-context
+        annotations:
+            config.kubernetes.io/local-config: "true"
+    spec:
+        cniConfig:
+            cniType: macvlan
+            masterInterface: eth1
+        siteCode: edge1
+        region: us-central1
+
+the following list of CRs will be added to package:
+
+    apiVersion: ipam.nephio.org/v1alpha1
+    kind: IPAllocation
+    metadata:
+      name: n6-ip-abcd
+      annotations:
+        krm-fn.nephio.org/owner: req.nephio.org/v1alpha1/interface/n6
+    spec:
+      kind: network
+      prefixLength: 32
+      networkInstanceRef: 
+        namespace: default
+        name: sample-vpc
+      selector:
+        matchLabels:
+          nephio.org/region: us-central1
+          nephio.org/site: edge1
+          nephio.org/network-name: net1   # TODO: Does this come from ClusterContext?
+    ---
+    apiVersion: ipam.nephio.org/v1alpha1
+    kind: VLANAllocation
+    metadata:
+      name: n6-vlan-defg
+      annotations:
+        krm-fn.nephio.org/owner: req.nephio.org/v1alpha1/interface/n6
+    spec:
+      networkInstanceRef: 
+        namespace: default
+        name: sample-vpc
+      selector:
+        matchLabels:
+          nephio.org/region: us-central1
+          nephio.org/site: edge1
+          nephio.org/network-name: net1
+
+
+The function also adds the following conditions to the `Status` of the `Kptfile` object:
+
+    apiVersion: kpt.dev/v1
+    kind: Kptfile
+    metadata:
+        name: pkg-upf
+        annotations:
+            config.kubernetes.io/local-config: "true"
+    pipeline:
+      mutators:
+      - image: nephio.org/interface-manager:v1
+      - image: nephio.org/vlan:v1
+      - .....
+    status:
+        conditions:
+            - type: req-nephio-org-v1alpha1-interface-n6
+              status: true
+            - type: ipam-nephio-org-v1alpha1-ipallocation-n6-ip-abcd
+              status: false
+            - type: ipam-nephio-org-v1alpha1-vlanallocation-n6-vlan-defg
+              status: false
+            - type: req-nephio-org-v1alpha1-interface-n6-nad-generated
+              status: false
+
+
+Implementation details
+----------------------
+
```

Comment:
Once the function runtime library client API is decided, it would be good to mention how that would be used here as well.